### PR TITLE
RFC30/number

### PR DESCRIPTION
**Pull request by thomas-k-cameron**
## Motivation and Context
part of RFC30

## Description
add serde support for number
refactoring: move number and related module/functions to another file

## Testing
- de-serializes 0, positive integer, negative integer and number with component below the decimal point.
- serializes 0, positive integer, negative integer and number with component below the decimal point.

## Checklist
NA

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-types/src/lib.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-types/src/lib.rs:26]: Can you move this `use` statement down with the other `use` statements?

File: rust-runtime/aws-smithy-types/src/number.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-types/src/number.rs:29]: @rcoh Is this a dev guide thing?

File: rust-runtime/aws-smithy-types/src/number.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-types/src/number.rs:458]: Let's bring this in line with your datetime PR
```suggestion
    #[test]
    #[cfg(all(test, aws_sdk_unstable, feature = "deserialize", feature = "serialize"))]
```

File: rust-runtime/aws-smithy-types/src/number.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-types/src/number.rs:29]: It is.

File: rust-runtime/aws-smithy-types/src/lib.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/lib.rs:26]: https://github.com/awslabs/smithy-rs/pull/2197

I made a draft PR for adding `rustfmt.toml`.   
`cargo fix` can automatically fix up things like the one you've mentioned and I think these things should be automated.

Can you check my PR?

Comment (Velfi) [reply rust-runtime/aws-smithy-types/src/lib.rs:26]: Great idea, I'll check it out.

**Comment by thomas-k-cameron** [comment]
fixed
https://github.com/awslabs/smithy-rs/pull/2185#discussion_r1065971687

**Comment by Velfi** [comment]
Hey @thomas-k-cameron, I broke a datetime import when merging your other PR. Sorry about that. I fixed it but I can't merge it into your branch. Here's the branch with the fix: https://github.com/awslabs/smithy-rs/tree/fix/datetime-missing-import

**Comment by thomas-k-cameron** [comment]
https://github.com/awslabs/smithy-rs/pull/2185#issuecomment-1397608116

No problem!
let me fix it up.

I couldn't really make any commits this week as I am kinda busy. I should be able to start picking up from this weekend.

**Comment by thomas-k-cameron** [comment]
@Velfi 
should be fixed

File: codegen-server-test/python/model/pokemon-common.smithy
Comment (jdisanti) [inline codegen-server-test/python/model/pokemon-common.smithy:1]: What changed with this symlink? Its showing up as both deleted and created in the GitHub diff.

File: rust-runtime/aws-smithy-types/src/lib.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/lib.rs:19]: What is this change for?

**Review by jdisanti — COMMENTED** [review]
It looks like the changes in this PR got a bit muddied by merging various other branches into it. I think you should start from a clean slate and pull in the desired number changes again to clean things up.

File: rust-runtime/aws-smithy-types/src/lib.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/lib.rs]: It causes a import error without it.

If you check this file(smithy-rs/rust-runtime/aws-smithy-types/src/date_time/format.rs), you can find 5 of `use crate::DateTime`.

mod.rs file has it one too.

I'm not quite sure how it ended up this way though.

File: codegen-server-test/python/model/pokemon-common.smithy
Comment (thomas-k-cameron) [reply codegen-server-test/python/model/pokemon-common.smithy]: I have no idea :(

Comment (thomas-k-cameron) [reply codegen-server-test/python/model/pokemon-common.smithy]: I reverted the commit that was causing this. It should work now.

File: rust-runtime/aws-smithy-types/src/lib.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/lib.rs:29]: Looks like the `DateTime` and `Error` re-exports were lost:
```suggestion
pub use blob::Blob;
pub use date_time::DateTime;
pub use document::Document;
pub use error::Error;
pub use number::Number;
```

Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/lib.rs:29]: Sorry I didn't notice that.  
It seems like it got lost when I merged the other branch.  

Thank you for letting me know!

**Comment by jdisanti** [comment]
It looks like there are just some unused imports that need to be fixed to make CI pass, and then I will be ready to approve and merge.

**Comment by thomas-k-cameron** [comment]
Can someone run the test for me?

**Comment by thomas-k-cameron** [comment]
Weird error. Let me look into it.

**Comment by thomas-k-cameron** [comment]
Ok I think it works now.

**Comment by thomas-k-cameron** [comment]
@jdisanti 
Would you kindly run the CI for me?

**Comment by thomas-k-cameron** [comment]
@jdisanti 

Thank you for running the CI.

I noticed that I got this error; I'm not quite sure what is this about but I'm suspecting that I could fix this up by merging the main branch.
```
> Task :aws:sdk:generateVersionManifest
2023-01-30T23:42:07.786751Z  INFO publisher::subcommand::generate_version_manifest: Resolved smithy-rs revision to 716bfd436c1036cc450a3d96b07be838e82135d2
crate-hasher 0.1.0
2023-01-30T23:43:16.151920Z  INFO publisher::subcommand::generate_version_manifest: Discovered and hashed 355 crates
Error: Version number for `aws-config` decreased between releases (from `0.54.1` to `0.54.0`)

> Task :aws:sdk:generateVersionManifest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':aws:sdk:generateVersionManifest'.
39 actionable tasks: 38 executed, 1 up-to-date
> Process 'command 'publisher'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org/
```

**Comment by thomas-k-cameron** [comment]
OK, I merged the main branch.

**Comment by jdisanti** [comment]
> I noticed that I got this error; I'm not quite sure what is this about but I'm suspecting that I could fix this up by merging the main branch.

Yeah, I've run into this one before 😅 Merging main fixes it. Fortunately, this version check will be getting removed in https://github.com/awslabs/smithy-rs/pull/2253

**Comment by thomas-k-cameron** [comment]
I merged it but I started running into other issues so I decided to revert it 👯‍♂️

**Comment by thomas-k-cameron** [comment]
Considering that the only error is 
```
> Task :aws:sdk:generateVersionManifest
```
I think it's ready to be merged.

**Comment by thomas-k-cameron** [comment]
@Velfi @jdisanti 
Can you merge this?

**Comment by jdisanti** [comment]
I created https://github.com/awslabs/smithy-rs/pull/2303 to update the `unstable-serde-support` branch with main, which will allow us to update this branch so we can see CI passing. I'll merge this prerequisite PR after confirming the tests pass.

**Comment by thomas-k-cameron** [comment]
Thank you!

**Comment by jdisanti** [comment]
Something has gotten weird in the git history for this PR, and I don't see an easy way to fix it. It seems every time we merge `unstable-serde-support` into it, a bunch of unrelated changes appear.

@thomas-k-cameron - Can you copy your changed `aws-smithy-types` code somewhere, git reset (hard) the history on this PR to the latest `unstable-serde-support` branch, copy the changes back in, commit, and then force push?

**Comment by thomas-k-cameron** [comment]
Wait what

**Comment by jdisanti** [comment]
Looks like GitHub automatically closed it since the branch is identical to `unstable-serde-support`. I think you can reopen it if you add your commit now.

**Comment by thomas-k-cameron** [comment]
I think I kinda messed it up on it's way but it should have all the changes that it had before.

**Review by jdisanti — APPROVED** [review]

**Pull request merged**
